File: archive/archived_apps/ios_legacy_20250910/Backend/Python/tools/cash_tools.py
"""
AFL Fantasy Cash Tools

This module provides tools for analyzing cash generation and managing
rookie/low-price players in your fantasy team.
"""

from scraper import get_player_data


# Tool 1: Cash Generation Tracker
def cash_generation_tracker():
    """
    Track potential cash generation from players
    
    Returns:
        list: List of players with price change estimates
    """
    data = get_player_data()
    return [
        {
            "player": p["name"],
            "team": p["team"],
            "price": p["price"],
            "breakeven": p["breakeven"],
            "3_game_avg": p["l3_avg"],
            "price_change_est": round((p["l3_avg"] - p["breakeven"]) * 150),  # Scaled estimate
        }
        for p in data if p["games"] >= 2
    ]


# Tool 2: Rookie Price Curve Model
def rookie_price_curve_model():
    """
    Model the price curve for rookies
    
    Returns:
        list: List of rookies with price projections
    """
    data = get_player_data()
    rookies = [p for p in data if p["price"] < 500000 and p["games"] >= 2]
    return [
        {
            "player": r["name"],
            "price": r["price"],
            "l3_avg": r["l3_avg"],
            "price_projection_next_3": round(r["price"] + ((r["l3_avg"] - r["breakeven"]) * 150 * 3))
        }
        for r in rookies
    ]


# Tool 3: Downgrade Target Finder
def downgrade_target_finder():
    """
    Find potential downgrade targets based on rookies with low breakevens
    
    Returns:
        list: List of rookies with low breakevens, sorted by breakeven
    """
    data = get_player_data()
    return sorted(
        [p for p in data if p["price"] < 500000 and p["breakeven"] < 40],
        key=lambda x: x["breakeven"]
    )


# Tool 4: Cash Gen Ceiling/Floor
def cash_gen_ceiling_floor():
    """
    Calculate ceiling and floor price changes for players
    
    Returns:
        list: List of players with ceiling and floor price changes
    """
    data = get_player_data()
    result = []
    for p in data:
        if "l3_avg" in p and p["l3_avg"] > 0 and "breakeven" in p:
            diff = p["l3_avg"] - p["breakeven"]
            est_change = round(diff * 150)
            result.append({
                "player": p["name"],
                "team": p["team"],
                "price": p["price"],
                "floor": est_change - 3000,  # simulating bad games
                "ceiling": est_change + 3000,  # simulating breakout games
            })
    return result


# Tool 5: Price Predictor Calculator
def price_predictor_calculator(player_name=None, scores=None):
    """
    Calculate predicted price based on future scores
    
    Parameters:
        player_name (str, optional): Name of the player to predict price for
        scores (list, optional): List of predicted scores for next rounds
    
    Returns:
        dict: Price prediction data
    """
    if not player_name or not scores:
        return {"error": "Player name and scores are required"}
    
    data = get_player_data()
    
    # Find the player
    player = None
    for p in data:
        if p["name"].lower() == player_name.lower():
            player = p
            break
    
    if not player:
        return {"error": f"Player '{player_name}' not found"}
    
    # Calculate the price changes
    current_price = player["price"]
    breakeven = player["breakeven"]
    
    price_changes = []
    
    for i, score in enumerate(scores):
        # Calculate price change
        diff = score - breakeven
        price_change = round(diff * 150)
        new_price = current_price + price_change
        
        price_changes.append({
            "round": i + 1,
            "score": score,
            "price_change": price_change,
            "new_price": new_price
        })
        
        # Update for next round
        current_price = new_price
        # Adjust breakeven (simplified model)
        breakeven = round(breakeven + (diff * 0.1))
    
    return {
        "player": player_name,
        "starting_price": player["price"],
        "starting_breakeven": player["breakeven"],
        "price_changes": price_changes,
        "final_price": price_changes[-1]["new_price"] if price_changes else player["price"]
    }


# Tool 6: Price Ceiling/Floor Estimator
def price_ceiling_floor_estimator():
    """
    Estimate ceiling and floor prices for players
    
    Returns:
        list: List of players with ceiling and floor price estimates
    """
    data = get_player_data()
    result = []
    
    for p in data:
        if p["games"] >= 3:  # Only include players with enough games
            current_price = p["price"]
            avg = p["avg"]
            be = p["breakeven"]
            
            # Calculate ceiling (assuming 25% above average)
            ceiling_avg = avg * 1.25
            ceiling_diff = ceiling_avg - be
            ceiling_price = current_price + round(ceiling_diff * 150 * 6)  # Project 6 rounds ahead
            
            # Calculate floor (assuming 25% below average)
            floor_avg = avg * 0.75
            floor_diff = floor_avg - be
            floor_price = current_price + round(floor_diff * 150 * 6)  # Project 6 rounds ahead
            
            # Cap the floor price at minimum
            floor_price = max(floor_price, 102000)  # Minimum player price
            
            result.append({
                "player": p["name"],
                "team": p["team"],
                "position": p["position"],
                "current_price": current_price,
                "ceiling_price": ceiling_price,
                "floor_price": floor_price,
                "ceiling_gain": ceiling_price - current_price,
                "floor_loss": current_price - floor_price
            })
    
    # Sort by ceiling gain (highest first)
    return sorted(result, key=lambda x: x["ceiling_gain"], reverse=True)


if __name__ == "__main__":
    # Simple test to show the cash generation tracker when run directly
    print("Cash Generation Tracker:")
    cash_gen = cash_generation_tracker()
    # Show the top 5 players by price change estimate
    for player in sorted(cash_gen, key=lambda x: x["price_change_est"], reverse=True)[:5]:
        print(f"{player['player']} (${player['price']:,}): BE {player['breakeven']}, " +
              f"3-Game Avg {player['3_game_avg']}, Est. Change ${player['price_change_est']:+,}")
    
    print("\nDowngrade Target Finder:")
    # Show the top 5 downgrade targets
    for player in downgrade_target_finder()[:5]:
        print(f"{player['name']} (${player['price']:,}): BE {player['breakeven']}")

File: archive/server-node-backup/backend/cash-tools.ts
/**
 * AFL Fantasy Cash Generation Tools
 */
import { DEFAULT_MAGIC_NUMBER, calculatePriceChange } from './utils';
import { FantasyPlayer } from './trade-tools';

/**
 * Track projected cash generation for a player
 * 
 * @param player Player to analyze
 * @param projectedScores Array of projected scores
 * @param magicNumber Magic number for price calculations
 * @returns Analysis of projected cash generation
 */
export function trackCashGeneration(
  player: FantasyPlayer,
  projectedScores?: number[],
  magicNumber: number = DEFAULT_MAGIC_NUMBER
): {
  player: FantasyPlayer,
  currentPrice: number,
  projectedPrices: number[],
  totalCashGenerated: number,
  averageChange: number,
  cashGenerationRate: string,
  recommendation: string
} {
  // If no projected scores provided, use the player's projected score
  const scores = projectedScores || Array(5).fill(player.projectedScore);
  
  // Simulate price changes over the next few rounds
  let currentPrice = player.price;
  let currentBreakEven = player.breakeven;
  const projectedPrices = [currentPrice];
  
  // For each projected score, calculate price change
  for (let i = 0; i < scores.length; i++) {
    const score = scores[i];
    const priceChange = calculatePriceChange(score, currentBreakEven, magicNumber);
    currentPrice += priceChange;
    projectedPrices.push(currentPrice);
    
    // Update breakeven for next round (simplified approximation)
    currentBreakEven = Math.round((currentPrice / magicNumber) * 0.9);
  }
  
  // Calculate total cash generated
  const totalCashGenerated = projectedPrices[projectedPrices.length - 1] - player.price;
  
  // Calculate average weekly change
  const averageChange = totalCashGenerated / scores.length;
  
  // Determine cash generation rate
  let cashGenerationRate = "neutral";
  if (averageChange > 10000) cashGenerationRate = "high";
  else if (averageChange > 5000) cashGenerationRate = "medium";
  else if (averageChange > 0) cashGenerationRate = "low";
  else cashGenerationRate = "negative";
  
  // Generate recommendation
  let recommendation = "";
  if (cashGenerationRate === "high") {
    recommendation = "Strong cash cow, hold until price peaks";
  } else if (cashGenerationRate === "medium") {
    recommendation = "Moderate cash generation, consider holding for 2-3 more rounds";
  } else if (cashGenerationRate === "low") {
    recommendation = "Limited upside, trade when a better option is available";
  } else {
    recommendation = "Negative cash generation, trade out soon";
  }
  
  return {
    player,
    currentPrice: player.price,
    projectedPrices,
    totalCashGenerated,
    averageChange,
    cashGenerationRate,
    recommendation
  };
}

/**
 * Model a rookie's price growth based on scoring progression
 * 
 * @param player Rookie player to model
 * @param scores Historical scores
 * @param projectedScores Projected future scores
 * @param initialPrice Initial price (defaults to current price if not provided)
 * @param magicNumber Magic number for price calculations
 * @returns Analysis of rookie price curve
 */
export function modelRookiePriceCurve(
  player: FantasyPlayer,
  scores: number[],
  projectedScores: number[],
  initialPrice?: number,
  magicNumber: number = DEFAULT_MAGIC_NUMBER
): {
  player: FantasyPlayer,
  initialPrice: number,
  currentPrice: number,
  projectedPeakPrice: number,
  projectedPeakRound: number,
  priceTrajectory: number[],
  totalGrowth: number,
  growthPercentage: number,
  recommendation: string
} {
  // Use provided initial price or default to current price
  const startPrice = initialPrice || player.price;
  
  // Combine historical and projected scores
  const allScores = [...scores, ...projectedScores];
  
  // Initialize price trajectory with initial price
  const priceTrajectory = [startPrice];
  let currentPrice = startPrice;
  let currentBE = player.breakeven;
  let peakPrice = startPrice;
  let peakRound = 0;
  
  // Simulate price changes based on AFL Fantasy algorithm
  // For rookies, prices change after 3 games based on average
  if (scores.length < 3) {
    // Rookie hasn't played 3 games yet
    // AFL Fantasy doesn't change prices until 3 games are played
    for (let i = scores.length; i < 3 && i < allScores.length; i++) {
      priceTrajectory.push(currentPrice); // No change yet
    }
    
    // After 3 games, price is based on 3-game average
    if (allScores.length >= 3) {
      const first3Avg = (allScores[0] + allScores[1] + allScores[2]) / 3;
      currentPrice = Math.round(first3Avg * magicNumber);
      priceTrajectory.push(currentPrice);
      
      if (currentPrice > peakPrice) {
        peakPrice = currentPrice;
        peakRound = 3;
      }
      
      // Calculate new breakeven
      currentBE = Math.round((currentPrice / magicNumber) * 0.9);
      
      // Process remaining scores
      for (let i = 3; i < allScores.length; i++) {
        const score = allScores[i];
        const priceChange = calculatePriceChange(score, currentBE, magicNumber);
        currentPrice += priceChange;
        priceTrajectory.push(currentPrice);
        
        // Update peak price
        if (currentPrice > peakPrice) {
          peakPrice = currentPrice;
          peakRound = i;
        }
        
        // Update breakeven for next round
        currentBE = Math.round((currentPrice / magicNumber) * 0.9);
      }
    }
  } else {
    // Rookie has already played 3+ games
    // Process all projected scores
    for (let i = 0; i < projectedScores.length; i++) {
      const score = projectedScores[i];
      const priceChange = calculatePriceChange(score, currentBE, magicNumber);
      currentPrice += priceChange;
      priceTrajectory.push(currentPrice);
      
      // Update peak price
      if (currentPrice > peakPrice) {
        peakPrice = currentPrice;
        peakRound = scores.length + i;
      }
      
      // Update breakeven for next round
      currentBE = Math.round((currentPrice / magicNumber) * 0.9);
    }
  }
  
  // Calculate total growth and percentage
  const totalGrowth = peakPrice - startPrice;
  const growthPercentage = (totalGrowth / startPrice) * 100;
  
  // Generate recommendation
  let recommendation = "";
  if (player.price >= peakPrice) {
    recommendation = "Rookie has likely peaked, consider trading out now";
  } else if (peakRound - scores.length <= 1) {
    recommendation = "Rookie will peak very soon, prepare trade plans";
  } else if (peakRound - scores.length <= 3) {
    recommendation = "Rookie will peak in next 2-3 rounds, hold for now";
  } else {
    recommendation = "Rookie has significant growth potential, strong hold";
  }
  
  return {
    player,
    initialPrice: startPrice,
    currentPrice: player.price,
    projectedPeakPrice: peakPrice,
    projectedPeakRound: peakRound,
    priceTrajectory,
    totalGrowth,
    growthPercentage,
    recommendation
  };
}

/**
 * Find optimal downgrade targets with good value
 * 
 * @param allPlayers All available players
 * @param currentTeam Current team players
 * @param maxPrice Maximum price for downgrade targets
 * @param minProjectedScore Minimum projected score to consider
 * @returns List of downgrade targets ranked by value
 */
export function findDowngradeTargets(
  allPlayers: FantasyPlayer[],
  currentTeam: FantasyPlayer[],
  maxPrice: number = 300000,
  minProjectedScore: number = 65
): {
  player: FantasyPlayer,
  valueIndex: number,
  projectedScoreLoss: number,
  cashSaved: number,
  namedStatus: string
}[] {
  // Filter to players not in current team, below max price, with adequate projected scoring
  const downgradeTargets = allPlayers.filter(
    p => !currentTeam.some(tp => tp.id === p.id) &&
         p.price <= maxPrice &&
         p.projectedScore >= minProjectedScore
  );
  
  // Calculate value index for each target
  const targets = downgradeTargets.map(player => {
    // Calculate points per $1000
    const valueIndex = (player.projectedScore * 1000) / player.price;
    
    // Find comparable player in same position on current team
    const teamPlayerSamePos = currentTeam
      .filter(p => p.position === player.position)
      .sort((a, b) => a.price - b.price)[0]; // Get cheapest as comparison
    
    // Calculate projected score loss vs cheapest player in same position
    const projectedScoreLoss = teamPlayerSamePos ? 
                               teamPlayerSamePos.projectedScore - player.projectedScore :
                               0;
    
    // Calculate cash saved vs player being replaced
    const cashSaved = teamPlayerSamePos ? 
                      teamPlayerSamePos.price - player.price :
                      0;
    
    // Determine if player is named in their team (simplified approximation)
    // In a real implementation, this would check actual team selection data
    const namedStatus = player.redDotFlag ? "Injured/Suspended" : "Named";
    
    return {
      player,
      valueIndex,
      projectedScoreLoss,
      cashSaved,
      namedStatus
    };
  });
  
  // Sort by value index (descending)
  return targets.sort((a, b) => b.valueIndex - a.valueIndex);
}

/**
 * Estimate the floor and ceiling of a player's price
 * 
 * @param player Player to analyze
 * @param floorScore Estimated floor score
 * @param ceilingScore Estimated ceiling score
 * @param magicNumber Magic number for price calculations
 * @returns Analysis of player's price range
 */
export function calculatePriceRange(
  player: FantasyPlayer,
  floorScore?: number,
  ceilingScore?: number,
  magicNumber: number = DEFAULT_MAGIC_NUMBER
): {
  player: FantasyPlayer,
  currentPrice: number,
  floorScore: number,
  ceilingScore: number,
  floorPrice: number,
  ceilingPrice: number,
  projectedEndPrice: number,
  recommendation: string
} {
  // If floor/ceiling not provided, estimate them
  const floor = floorScore || 
                (player.floor || 
                (player.projectedScore ? Math.max(player.projectedScore - 20, 40) : 40));
  
  const ceiling = ceilingScore || 
                  (player.ceiling || 
                  (player.projectedScore ? player.projectedScore + 20 : 100));
  
  // Calculate price floor - price if player scores at floor for 3 consecutive games
  const floorAvg = floor;
  const floorPrice = Math.round(floorAvg * magicNumber);
  
  // Calculate price ceiling - price if player scores at ceiling for 3 consecutive games
  const ceilingAvg = ceiling;
  const ceilingPrice = Math.round(ceilingAvg * magicNumber);
  
  // Calculate projected end price based on player's projected score
  const projectedAvg = player.projectedScore || player.average || 
                      ((floor + ceiling) / 2);
  const projectedEndPrice = Math.round(projectedAvg * magicNumber);
  
  // Generate recommendation
  let recommendation = "";
  
  // If current price is close to floor
  if (player.price < floorPrice * 1.1) {
    recommendation = "Player priced near floor, good value to trade in";
  } 
  // If current price is close to ceiling
  else if (player.price > ceilingPrice * 0.9) {
    recommendation = "Player priced near ceiling, consider trading out";
  }
  // If projected price is higher than current price
  else if (projectedEndPrice > player.price * 1.1) {
    recommendation = "Player likely to increase in value";
  }
  // If projected price is lower than current price
  else if (projectedEndPrice < player.price * 0.9) {
    recommendation = "Player likely to decrease in value";
  }
  // Otherwise
  else {
    recommendation = "Player fairly priced, trading decision should be based on other factors";
  }
  
  return {
    player,
    currentPrice: player.price,
    floorScore: floor,
    ceilingScore: ceiling,
    floorPrice,
    ceilingPrice,
    projectedEndPrice,
    recommendation
  };
}

/**
 * Calculate cash generation ceiling and floor for a player
 * 
 * @param player Player to analyze
 * @param projectedScores Array of projected scores
 * @param magicNumber Magic number for price calculations
 * @returns Analysis of cash generation ceiling and floor
 */
export function calculateCashGenCeilingFloor(
  player: FantasyPlayer,
  projectedScores?: number[],
  magicNumber: number = DEFAULT_MAGIC_NUMBER
): {
  player: FantasyPlayer,
  currentPrice: number,
  floorPrice: number,
  ceilingPrice: number,
  breakEvenFloor: number,
  breakEvenCeiling: number,
  maxCashGeneration: number,
  minCashGeneration: number,
  recommendation: string
} {
  // Estimate score floor and ceiling if not available
  const floorScore = player.floor || 
                    (player.projectedScore ? player.projectedScore * 0.8 : 
                    (player.average ? player.average * 0.8 : 40));
  
  const ceilingScore = player.ceiling || 
                      (player.projectedScore ? player.projectedScore * 1.2 : 
                      (player.average ? player.average * 1.2 : 100));
  
  // Calculate price ceiling and floor
  const priceFloor = player.price;
  let priceCeiling = player.price;
  let breakEvenFloor = player.breakeven;
  let breakEvenCeiling = player.breakeven;
  
  // Simulate price changes for floor scenario (consistently poor scores)
  for (let i = 0; i < 3; i++) {
    const priceChange = calculatePriceChange(floorScore, breakEvenFloor, magicNumber);
    breakEvenFloor = Math.round((breakEvenFloor + priceChange) * 0.9);
  }
  const floorPrice = Math.max(priceFloor + calculatePriceChange(floorScore, breakEvenFloor, magicNumber), 0);
  
  // Simulate price changes for ceiling scenario (consistently high scores)
  for (let i = 0; i < 3; i++) {
    const priceChange = calculatePriceChange(ceilingScore, breakEvenCeiling, magicNumber);
    priceCeiling += priceChange;
    breakEvenCeiling = Math.round((priceCeiling / magicNumber) * 0.9);
  }
  const ceilingPrice = priceCeiling;
  
  // Calculate maximum and minimum cash generation
  const maxCashGeneration = ceilingPrice - player.price;
  const minCashGeneration = floorPrice - player.price;
  
  // Generate recommendation
  let recommendation = "";
  if (maxCashGeneration > 100000 && minCashGeneration > 0) {
    recommendation = "Strong cash cow, very likely to increase in value";
  } else if (maxCashGeneration > 50000 && minCashGeneration > -20000) {
    recommendation = "Moderate cash generation potential with low risk";
  } else if (maxCashGeneration > 0 && minCashGeneration < 0) {
    recommendation = "Volatile cash generation, monitor closely";
  } else {
    recommendation = "Likely to lose value, trade out soon";
  }
  
  return {
    player,
    currentPrice: player.price,
    floorPrice,
    ceilingPrice,
    breakEvenFloor,
    breakEvenCeiling,
    maxCashGeneration,
    minCashGeneration,
    recommendation
  };
}